nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/common/AbstractMQTTProcessor.java

```diff
@@ -0,0 +1,190 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.processors.mqtt.common;
+
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.ValidationContext;
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.components.Validator;
+import org.apache.nifi.expression.AttributeExpression;
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.ssl.SSLContextService;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Properties;
+
+public abstract class AbstractMQTTProcessor extends AbstractProcessor{
+
+    public static final int LAST_WILL_QOS = 2;
+
+    public static final Validator QoSValidator = new Validator() {
+
+        @Override
+        public ValidationResult validate(String subject, String input, ValidationContext context) {
+            Integer inputInt = Integer.parseInt(input);
+            if (inputInt < 0 || inputInt > 2) {
+                return new ValidationResult.Builder().subject(subject).valid(false).explanation("QoS must be an integer between 0 and 2").build();
+            }
+            return new ValidationResult.Builder().subject(subject).valid(true).build();
+        }
+    };
+
+    public static final Validator BrokerValidator = new Validator() {
+
+        @Override
+        public ValidationResult validate(String subject, String input, ValidationContext context) {
+            try{
+                URI brokerURI = new URI(input);
+                if (!"".equals(brokerURI.getPath())) {
+                    return new ValidationResult.Builder().subject(subject).valid(false).explanation("the broker URI cannot have a path. It currently is:" + brokerURI.getPath()).build();
+                }
+                if (!("tcp".equals(brokerURI.getScheme()) || "ssl".equals(brokerURI.getScheme()))) {
+                    return new ValidationResult.Builder().subject(subject).valid(false).explanation("only the 'tcp' and 'ssl' schemes are supported.").build();
+                }
+            } catch (URISyntaxException e) {
+                e.printStackTrace();
```

Comment:
https://45.media.tumblr.com/9f90cb40ef2580146399b7588a33c966/tumblr_o0ose9Noon1rc0hxao1_400.gif

What are you talking about ....?

lol but yeah that's my bad